### ISSUE TITLE
Malicious plugin author action support request.

### ISSUE THREAD

**Issue by cgarz** [BetterDiscord/BetterDiscord#1392]
Earlier today one of the plugins installed in my Better Discord forcibly and permanently disabled all its functionality and replaced itself with a message containing a dark pattern seemingly designed to manipulate users into deleting the file.

There was no prior indication or confirmation of these actions and as of now I am unable to restore the plugin to its previous state as all the code has been replaced.

The plugin in question is Show Hidden Channels by DevilBro.

The plugin was installed via the official Better Discord website. However, I understand that recently the plugin has had support dropped due to updated guidelines. Nonetheless it is clearly unacceptable to delete files from users computers without their consent and attempt to manipulate them with dark patterns.

As such I am inquiring how to raise this issue further so that action can be taken regarding these clearly malicious actions as per the code of conduct mentioned in the [CONTRIBUTING](https://github.com/BetterDiscord/BetterDiscord/blob/main/CONTRIBUTING.md) readme as well as the new updated guidelines themselves. Which I feel this plugin author has clearly went against both of.

I apologize if this is the wrong place to raise such an issue. I initially inquired on the Better Discord server and was eventually directed  here after being rudely dismissed and having to explain why such a thing was even a concern.

I also now understand after some reading that there has been some degree of friction in the Better Discord community on either side of this issue regarding the plugins removal. Despite that it is my hope that these concerns can still be addressed with civility here as opposed to the responses to the support request on the Better Discord server.

**Comment by qwerasd205** [comment]
> Nonetheless it is clearly unacceptable to delete files from users computers without their consent and attempt to manipulate them with dark patterns.
There was exactly as much consent involved as there is for any plugin update (which never happen automatically), if not more, thanks to the additional popup before a full delete of the file. BDFDB updates are only ever initiated by clicking to update in the notice bar at the top of the app.

DevilBro certainly could have handled the end-of-life update better, but it's a stretch to call it "malicious". If he had chosen to just let the plugin become stale and break over time because he stopped maintaining it then he would have received a flood of clueless users contacting him about it. Doing it this way allowed him to inform all of the users of the plugin that it will no longer be maintained in any official capacity, and if they choose to go seek out an unmaintained copy of the plugin then they (hopefully) won't come to him when it breaks.

To be malicious, the behaviour would have to be intended to do something harmful to users. A plugin which will no longer be maintained informing users of that fact and removing itself to avoid confusion is not a harmful action.

**Comment by cgarz** [comment]
> There was exactly as much consent involved as there is for any plugin update (which never happen automatically), if not more, thanks to the additional popup before a full delete of the file.

The situation here is very different from other plugin updates. This in essence is not an update at all. The plugin that is left after the "update" process is entirely different, and even explicitly lacks the feature that users install it for in the first place.

> EDIT: It has come to my attention the previous statement may have been inaccurate, which will be investigated further.

> DevilBro certainly could have handled the end-of-life update better, but it's a stretch to call it "malicious"

> if they choose to go seek out an unmaintained copy of the plugin then they (hopefully) won't come to him when it breaks.

Yes this is another issue that I have recently discovered when attempting to restore the file from a backup, it seems that BDFDB has also been changed to even go so far as to block users from restoring the file themselves from a backup. The restored file is instantly overwritten when loaded.

This is an even more serious and now hopefully very clearly malicious action as now DevilBro is actively trying to dictate what end users are able to achieve on their own machines even after his final update. End users are now unable to seek out and use an unmaintained copy as you suggest.

> To be malicious, the behaviour would have to be intended to do something harmful to users. A plugin which will no longer be maintained informing users of that fact and removing itself to avoid confusion is not a harmful action.

The update process completely and permanently disabling the plugin automatically is far beyond just stopping support. The user has decided that they want to have the feature and DevilBro has decided to automatically overrule that decision. That is malicious.
The message could easily have been added without breaking it. He could even have automatically disabled the plugin and ket the deletion message, so it was optional. DevilBro has very deliberately forced his decision on end users and has even now went so far as to create barriers to them manually restoring it themselves.

This is honestly extremely concerning behavior for a trusted developer on such a popular piece of software to be exhibiting, and doubly so is the lackluster response to his actions.

**Comment by Puv1s** [comment]
Quoting announcement from Zerebos:

It has come to the attention of the core BetterDiscord team that the end-of-life update for ShowHiddenChannels by DevilBro was automatically applied to users' plugin files without any prerequisite interaction or consent. The potential for this behavior in DevilBro's updater had gone under the radar until now because his library was grandfathered in before the new review system was introduced and assumed to be compliant when more thorough plugin review standards were originally enacted.
﻿
Unequivocally, this behavior is unacceptable and should not have occurred or have been possible in the first place. We apologize to anyone inconvenienced or otherwise perturbed by what happened.
﻿
DevilBro has been instructed to remove the forced update capability from his library as soon as possible and has agreed to make this change, however, additional measures will be taken to avoid a similar incident of any kind in the future. Specifically, we are working to deploy an update to BetterDiscord which features an integrated addon updater which will always respect user choice, and upon release we will immediately disallow 3rd-party updaters of any kind in officially listed plugins. This update has been in the works for some time now, and though we unfortunately can't travel back in time and release it a week ago to stop any of this from having happened, we'll do the next best thing and fast-track its release starting now.

**Comment by cgarz** [comment]
That is very refreshing to finally hear.

However what of the many users who have already been stung by this unacceptable behavior?

It would be ideal if an official copy of the original plugin could be posted in a trusted location so that anyone who wishes to restore what was forcibly removed from their computer without consent can do so. As opposed to finding potentially malicious copies from untrusted sources.

**Comment by qwerasd205** [comment]
Unfortunately since the plugin violates the new guidelines, it would be somewhat contradictory to provide it "officially". However, now that DevilBro has removed the forced up date logic from BDFDB, if you get a (currently) working copy of it from the history of his Git repo you can choose to ignore the update notices.

**Comment by cgarz** [comment]
That's unfortunate. I worry that many users unfamiliar with GitHub will see its removal from master and not realize they can get it from a previous commit. For example one commit before the malicious commit:
[b510da9b3ad72ba3b773f7d79e391a459f31732d/Plugins/ShowHiddenChannels](https://github.com/mwittrien/BetterDiscordAddons/blob/b510da9b3ad72ba3b773f7d79e391a459f31732d/Plugins/ShowHiddenChannels/ShowHiddenChannels.plugin.js)

This is especially compounded by the fact that simply discussing unsupported plugins or linking to even such an official link is forbidden on the support server now. For everyone, not even just mods and people directly involved with the project. Such information is quickly deleted, often additionally with further penalties too.

These factors will greatly increase the likelihood of end users potentially using malicious copies and as such I feel some form of an exception should be made due to this risk. Perhaps adding to the announcement that people should not trust other sources than old commits from the mwittrien repo for a source of an old copy if they are truly determined to restore it.

**Comment by yT7Af9Fc** [comment]
> Which leaves me to wonder if it is yet another malicious action to force users to be unable to use the plugin. Something which will only serve to further drive users to potentially malicious copies.

Yes, the only purpose of that change was to prevent users from using the plugin if they avoided the automatic update somehow.

![image](https://user-images.githubusercontent.com/86932864/188817890-9c403633-511e-4383-be08-2fa85c49bb16.png)

**Comment by dudebot** [comment]
As of version 2.5.9 of the DBFDB plugin, there is still room for this to happen again via how it downloads and verifies "backups" of the library. Disabling the @updateUrl has no effect because the requestLibraryData() function requires to run before loadLibrary(). Without downloading and verifying this "backup" data, the library is designed to break because the InternalData of the backup *must* match that of the current plugin.

https://github.com/mwittrien/BetterDiscordAddons/blob/7e0ee843fc0549b69f75044f85c5750b97f90b6a/Library/0BDFDB.plugin.js#L1091
https://github.com/mwittrien/BetterDiscordAddons/blob/7e0ee843fc0549b69f75044f85c5750b97f90b6a/Library/0BDFDB.plugin.js#L1109

**Comment by QbDesu** [comment]
Plugins break, whether the author actively discontinues them or the plugin just breaks over time doesn't really matter. It's not malicious, malicious would be stealing a user's account, destroying their system, stealing data, etc. Calling the update that broke SHC is malicious is quite a stretch even if the mechanism that triggered the update without definitely shouldn't exist and was a security threat - but it was removed in the mean time.

The data the library loads is data used for module resolution, css classes for styling, and so on. These are to avoid having to push code updates whenever classes, change or discord modules get moved around. Importantly it doesn't load any code that will be executed. 

With the removal of the "heavily outdated" plugin auto update mechanism it can no longer apply code updates without user interaction. Meaning if the plugin could no longer push **actually** malicious code (account stealers, etc.) to the user without at least requiring the user to click on the update banner to load it.
In the future updates will be handled by BetterDiscord itself and will load updates from the site instead of GitHub directly, so there wouldn't even be an update notice if the update was found to be malicious. But that update is still in the works as of now.

If you consider a plugin author making their own plugin unusable "malicious" - quite honestly - boohoo.

**Comment by cgarz** [comment]
No one has said that authors simply discontinuing their plugins is malicious. If it was just discontinuation, meaning simply that no more work was done to maintain the plugin, that would of course have been absolutely fine.

What actually happened was entirely different, even disregarding the forced automatic update without consent.
An update was pushed that completely broke the plugin and even went so far as to include a dark pattern to fully remove the file.
There are no provisions within BetterDiscord to review the code before an update is applied. Not even the change log can be viewed before applying one. When people see that there is an update to a plugin which has always had earnest updates in the past they are not going to go and download the file outside discord so that they can do a line by line diff of the plugin, especially for an author that is considered trusted. DevilBro knew this would be the case and still pushed the update, which is why it can be labeled malicious.

Additionally, even with the forced update code removed, deliberate changes have been made to BDFDB to disallow users from restoring the file from a backup and having it run. These changes do nothing other than break ShowHiddenChannels.
Discontinuation would mean no more work is done regarding the plugin. What is actually happening is work is being done in relation the the plugin, work with the express and sole purpose of breaking its functionality. That is clearly malicious.

In any case the Better Discord support server is more suited for such tribalism. In this issue I would now rather focus on discussing what can be done to help end users from potentially downloading malicious copies of the plugin due to the unfortunate set of circumstances that DevilBro has helped create.

**Comment by Inve1951** [comment]
> Additionally, even with the forced update code removed, deliberate changes have been made to BDFDB to disallow users from restoring the file from a backup and having it run. These changes do nothing other than break ShowHiddenChannels.
Discontinuation would mean no more work is done regarding the plugin. What is actually happening is work is being done in relation the the plugin, work with the express and sole purpose of breaking its functionality. That is clearly malicious.

No, it's his code, he can break it if he wants. BDFDB is not meant to be used by other authors.

> ... what can be done to help [prevent] end users from potentially downloading malicious copies of the plugin ...

It should be as simple as not downloading any copies. But people have free will, and free stupidity - we can't stop them.
The only thing that can be done, and is being practiced on our official channels, is to tell people to stick to plugins provided at https://betterdiscord.app/plugins, which no longer includes SHC.

**Comment by cgarz** [comment]
> Calling the update that broke SHC is malicious is quite a stretch `even if the mechanism that triggered the update without definitely shouldn't exist` and was a security threat

> No, it's his code, he can break it if he wants.

I have already said the Better Discord support server is more suited for such tribalism.

**Comment by Puv1s** [comment]
I feel like square addressed every concern you had, what else is an issue here?

**Comment by cgarz** [comment]
The last one that I raised. https://github.com/BetterDiscord/BetterDiscord/issues/1392#issuecomment-1238724404

Clearly users want SHC. They are going to download it. And since it has been admitted that it should never have been forcibly deleted without consent in the first place, I feel that perhaps an announcement mentioning that should they feel strongly enough to manually replace it, then an older commit from the official repo before the malicious change would be safest.

**Comment by Puv1s** [comment]
> The last one that I raised. https://github.com/BetterDiscord/BetterDiscord/issues/1392#issuecomment-1238724404
> 
> Clearly users want SHC. They are going to download it. And since it has been admitted that it should never have been forcibly deleted without consent in the first place, I feel that perhaps an announcement mentioning that should they feel strongly enough to manually replace it, then an older commit from the official repo before the malicious change would be safest.

That's not really possible to do as there's been updates to the library, and the plugin simply wouldn't work without modifying the code of both the old plugin and library, and even if modified, has a high chance to break again in short time. It wouldn't make things better.

**Comment by dudebot** [comment]
> I feel like square addressed every concern you had, what else is an issue here?

Since this is an unrelated repo to bdfdb, until the new update feature comes out, there’s nothing that really can nor should be be done. The only thing I will add in context of square’s upcoming solution is that in order to avoid being the same dark pattern is to have three additional elements: an “ignore this update” option, a link to the associated commit, and third party code reviews to officially supported BD plugin updates. The last one might be a stretch considering this isn’t some corporation and everyone is doing this on their free time, but considering that as long as BDFDB.PluginUtils.downloadUpdate still exists in his library (which it does), the same thing can happen.

**Comment by cgarz** [comment]
Yes I have noticed that `0BDFDB.data.json` is being modified without consent which breaks SHC again. So that's `BDFDB.PluginUtils.downloadUpdate`?

Strongly agree that more options for updates are sorely needed. At the absolute minimum the changelog should be visible before the update. Ideally also the code with simple differences. Or a direct link so that users can do this themselves less tediously than they have to do currently.

**Comment by Puv1s** [comment]
> > I feel like square addressed every concern you had, what else is an issue here?
> 
> Since this is an unrelated repo to bdfdb, until the new update feature comes out, there’s nothing that really can nor should be be done. The only thing I will add in context of square’s upcoming solution is that in order to avoid being the same dark pattern is to have three additional elements: an “ignore this update” option, a link to the associated commit, and third party code reviews to officially supported BD plugin updates. The last one might be a stretch considering this isn’t some corporation and everyone is doing this on their free time, but considering that as long as BDFDB.PluginUtils.downloadUpdate still exists in his library (which it does), the same thing can happen.

We will make sure the functionality that caused this will be non-existent in his library or plugins anymore.

As for the new updater, you can be sure it's in our best interest to make the user have control around the updates, as well as security features from us such as the manual commit reviews, hash checking, etc.

**Comment by rockbyo5** [comment]
> Plugins break, whether the author actively discontinues them or the plugin just breaks over time doesn't really matter. It's not malicious, malicious would be stealing a user's account, destroying their system, stealing data, etc. Calling the update that broke SHC is malicious is quite a stretch even if the mechanism that triggered the update without definitely shouldn't exist and was a security threat - but it was removed in the mean time.
This alone says everything that needs to be said. DevilBro's plugins are his own code, his DBFDB library is his to update and if letting users keep using his plugin now that it is against Discord' s TOS goes against his goals, wishes or morals, it is his right to make an update to his library so that future versions aren't compatible with said plugin, speeding up the process where the plugin will break if not maintained by someone else.

There is a sound argument to be made about not wanting to annoy Discord too much and not to go too much against their TOS when possible. BetterDiscord is already technically against Discord's TOS inherently by virtue of being a modified client. So far, Discord tolerates it since it hasn't caused them much trouble, but as soon as Discord gets a bit too annoyed, they can try and take down this project in some way, be it legal recourses or some form of discord update making it impossible (or extremely hard) to modify the client.

What will you do when, after a "Insert X bad event on numerous large Discord servers" happens, Discord starts using checksums, cryptography or other technique to make it nearly impossible for BetterDiscord to keep existing? Or worse, Discord does a ban wave against tens of thousands of better discord users? I hope seeing hidden channels will have been worth it.

I fail to see any issue left now that the auto-update feature is removed and the official updating feature is coming. Yes, auto-updating a plugin with breaking changes is bad, but with DevilBro's contribution to this community, i'm happy to forgive this mistake.

**Comment by TheSpaniard** [comment]
> There is a sound argument to be made about not wanting to annoy Discord too much and not to go too much against their TOS when possible. BetterDiscord is already technically against Discord's TOS inherently by virtue of being a modified client. So far, Discord tolerates it since it hasn't caused them much trouble, but as soon as Discord gets a bit too annoyed, they can try and take down this project in some way, be it legal recourses or some form of discord update making it impossible (or extremely hard) to modify the client.

I understand your position however I believe we are a bit past hoping Discord doesn't take notice or trying to operate in a grey area. All it takes at this point is someone who doesn't care enough about their presence on Discord to make this blow up significantly. Sure they get banned, but if they feel strongly enough maybe it's worth it.

>What will you do when, after a "Insert X bad event on numerous large Discord servers" happens, Discord starts using checksums, cryptography or other technique to make it nearly impossible for BetterDiscord to keep existing? Or worse, Discord does a ban wave against tens of thousands of better discord users? I hope seeing hidden channels will have been worth it.

X bad thing, like someone changing files on a user's computer without any prompting or consent? If you wish to call it a mistake, it is a very costly one that can sink the entire project because Devilbro decided he didn't want people to make the "wrong" choice. Realistically, he has successful test data should he want to do this again. What is the expectation being set then? "Devilbro has written the majority of the plugins we have, and he decided to skip the user in taking an action because his "hand was forced". He won't do that again if he finds reason enough to do so...maybe, trust me bro." Yes it is his code, but everyone's consequences when he chooses to do what he did.

**Comment by joyobooto** [comment]
> X bad thing, like someone changing files on a user's computer without any prompting or consent? If you wish to call it a mistake, it is a very costly one that can sink the entire project because Devilbro decided he didn't want people to make the "wrong" choice. Realistically, he has successful test data should he want to do this again. What is the expectation being set then? "Devilbro has written the majority of the plugins we have, and he decided to skip the user in taking an action because his "hand was forced". He won't do that again if he finds reason enough to do so...maybe, trust me bro." Yes it is his code, but everyone's consequences when he chooses to do what he did.

Crazy how people don't see this as an issue and the guy hasn't been blacklisted. Literally installed it for a week and came to see wtf was happening, only to come across such weak responses from contributors... Good luck with the project I guess, don't see the need to take risks for just a few extra Discord functionalities.

**Comment by rauenzi** [comment]
As for the original issue here, we have indeed taken action against this developer, and also forced them to remove the code that allows them to force updates like they did. Since this has been removed from their library, every subsequent update from them will be given extra scrutiny during the manual review.

To be clear though, the BetterDiscord team stands behind the removal of the plugin from our website. However, we do not condone the subsequent actions of the developer that force removed the plugin without user approval.

As for the discussion around updates, the first version of the official update system has now been released and only updates through manually approved releases. I agree it would be a good utility to be able to link users to a diff or have a changelog that displays beforehand. If you want, feel free to open a new issue for that since it is out of scope for this one.
